sway/desktop/launcher.c

```diff
@@ -0,0 +1,210 @@
+		ws = ctx->node->sway_container->pending.workspace;
+		break;
+	case N_WORKSPACE:
+		ws = ctx->node->sway_workspace;
+		break;
+	case N_OUTPUT:
+		output = ctx->node->sway_output;
+		ws = workspace_by_name(ctx->name);
+		if (!ws) {
+			sway_log(SWAY_DEBUG,
+					"Creating workspace %s for pid %d because it disappeared",
+					ctx->name, ctx->pid);
+			if (!output->enabled) {
+				sway_log(SWAY_DEBUG,
+						"Workspace output %s is disabled, trying another one",
+						output->wlr_output->name);
+				output = NULL;
+			}
+			ws = workspace_create(output, ctx->name);
+		}
+		break;
+	case N_ROOT:
+		ws = workspace_create(NULL, ctx->name);
+		break;
+	}
+
+	return ws;
+}
+
+static void ctx_handle_node_destroy(struct wl_listener *listener, void *data) {
+	struct launcher_ctx *ctx = wl_container_of(listener, ctx, node_destroy);
+	switch (ctx->node->type) {
+	case N_CONTAINER:
+		// Unimplemented
+		break;
+	case N_WORKSPACE:;
+		struct sway_workspace *ws = ctx->node->sway_workspace;
+		wl_list_remove(&ctx->node_destroy.link);
+		wl_list_init(&ctx->node_destroy.link);
+		// We want to save this ws name to recreate later, hopefully on the
+		// same output
+		free(ctx->name);
+		ctx->name = strdup(ws->name);
+		if (!ws->output || ws->output->node.destroying) {
+			// If the output is being destroyed it would be pointless to track
+			// If the output is being disabled, we'll find out if it's still
+			// disabled when we try to match it.
+			ctx->node = &root->node;
+			break;
+		}
+		ctx->node = &ws->output->node;
+		wl_signal_add(&ctx->node->events.destroy, &ctx->node_destroy);
+		break;
+	case N_OUTPUT:
+		wl_list_remove(&ctx->node_destroy.link);
+		wl_list_init(&ctx->node_destroy.link);
+		// We'll make the ws ctx->name somewhere else
+		ctx->node = &root->node;
+		break;
+	case N_ROOT:
+		// Unreachable
+		break;
+	}
+}
+
+static void token_handle_destroy(struct wl_listener *listener, void *data) {
+	struct launcher_ctx *ctx = wl_container_of(listener, ctx, token_destroy);
+	ctx->token = NULL;
+	launcher_ctx_destroy(ctx);
+}
+
+struct launcher_ctx *launcher_ctx_create() {
+	struct sway_seat *seat = input_manager_current_seat();
+	struct sway_workspace *ws = seat_get_focused_workspace(seat);
+	if (!ws) {
+		sway_log(SWAY_DEBUG, "Failed to create launch context. No workspace.");
+		return NULL;
+	}
+
+	struct launcher_ctx *ctx = calloc(1, sizeof(struct launcher_ctx));
+	struct wlr_xdg_activation_token_v1 *token =
+		wlr_xdg_activation_token_v1_create(server.xdg_activation_v1);
+	token->data = ctx;
+	ctx->name = strdup(ws->name);
+	ctx->token = token;
+	ctx->node = &ws->node;
+
+	ctx->node_destroy.notify = ctx_handle_node_destroy;
+	wl_signal_add(&ctx->node->events.destroy, &ctx->node_destroy);
+
+	ctx->token_destroy.notify = token_handle_destroy;
+	wl_signal_add(&token->events.destroy, &ctx->token_destroy);
+
+	wl_list_init(&ctx->link);
+	return ctx;
+}
+
+const char *launcher_ctx_get_name(struct launcher_ctx *ctx) {
```

Comment:
Nit: maybe `get_token_name` would be slightly better.

Reply:
done.